### ISSUE TITLE
Keras `__init__.py` structure isn't readable by a static type checker

### ISSUE THREAD

**Issue by rchiodo** [keras-team/keras#19779]
In keras 3.2.1, if you import say keras.models, you get no errors in VS code:

![image](https://github.com/keras-team/keras/assets/19672699/21a37f87-b742-4222-9486-a8c6baa79c9d)

In keras 3.3.1, this same code shows an error:

![image](https://github.com/keras-team/keras/assets/19672699/475ef543-7220-4d66-85d3-8bf5ad618612)

I believe the reason for this was the refactoring of the `__init__.py` that happened in this commit:

https://github.com/keras-team/keras/commit/04891e89da87c2a433beb12ff7dad59403c71671

The code added in this commit dynamically adds the path to `api` to the sys path. This isn't something a static type checker can handle. 

Could this change be reverted? Or perhaps marked with a `if TYPECHECKING` special case?

**Comment by rchiodo** [comment]
Sorry I should have said, the code runs fine, it just can't be interpreted by a static type checker. Our users are confused by this because they can execute their code just fine, but Pylance doesn't think there's a module called `keras.models` and so all intellisense after that point doesn't work.

**Comment by fchollet** [comment]
@mattdangerw @sampathweb can we resolve this?

**Comment by fchollet** [comment]
> The code added in this commit dynamically adds the path to api to the sys path. This isn't something a static type checker can handle.

Basically, publicly-exported APIs are in `keras/api/`, rather than in the top-level `keras/` folder. This is not unlike, e.g. TF APIs, which live in `tf._api.v2`. Can PyLance only handle packages where the directory structure matches the public namespace?

**Comment by rchiodo** [comment]
> Can PyLance only handle packages where the directory structure matches the public namespace?

We should be able to handle anything that's through static imports. The old code was importing all of the api modules, so Pylance could tell that those were imported in the `__init__.py`.

The code in the commit I referenced changed to add the api path to the sys.path at runtime. We don't execute any python code, so we can't actually figure out that api should be on the sys.path.

It's kind of similar to how we don't support [PEP 660](https://peps.python.org/pep-0660/) because we can't execute the dynamic code in the `.pth` files. 

Here's a similar example:
https://github.com/microsoft/pyright/issues/6272

**Comment by rchiodo** [comment]
We also have a [problem ](https://github.com/microsoft/pylance-release/issues/3753) with tensorflow too when it imports keras like this:

```
from tensorflow.keras import activations, initializers, layers, losses, metrics, models, optimizers, regularizers
```

This is because keras is added to the tensorflow package dynamically.

**Comment by mattdangerw** [comment]
@rchiodo can you say more about the `if TYPECHECKING` special case?

**Comment by rchiodo** [comment]
I'd be happy to. 

If you've never seen it before, it's described here:
https://typing.readthedocs.io/en/latest/spec/directives.html#type-checking

It basically tells type checkers extra information that the runtime doesn't care about. 

So you might for example have this in your `__init__.py`

```python

if typing.TYPE_CHECKING:
  from keras.api import _tf_keras
  from keras.api import activations
  from keras.api import applications
  from keras.api import backend
  from keras.api import callbacks
  from keras.api import config
  from keras.api import constraint
  # etc
else:
   # Import everything from /api/ into keras.
   from keras.api import *  # noqa: F403
   from keras.api import __version__  # Import * ignores names start with "_".

   # Add everything in /api/ to the module search path.
   __path__.append(os.path.join(os.path.dirname(__file__), "api"))  # noqa: F405
```

**Comment by mattdangerw** [comment]
Thanks!

I guess if the issue is just with `import *`, we can probably just make sure we individually import everything from api -> toplevel. I don't think this needs to be done with `if TYPE_CHECKING` at all. We could even extend our api gen script to generate that block in the `__init__.py` file. @rchiodo @fchollet does that sounds like a plan?

**Comment by fchollet** [comment]
Sounds like a plan! This is clean solution.

**Comment by rchiodo** [comment]
Sorry but I don't know if there's a way to resolve this. Please see the comment I left in the PR:
https://github.com/keras-team/keras/pull/19864#issuecomment-2174204607

**Comment by Splines** [comment]
Does that mean that imports like the following are officially not supported by keras/tensorflow? In VSCode, something like this is not working for me. Both Pylint and Pylance complain in a virtual environment (`venv`). Other libraries like `matplotlib` work fine.

```py
import keras

from keras.datasets import cifar10
from keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout, Input
from keras.optimizers import SGD

import tensorflow as tf
```

![image](https://github.com/user-attachments/assets/8f301baf-ec53-464b-814b-4bcdf48295c3)

What is the recommended workaround for this? It's a pretty big deal for me not having autocompletion available available in VSCode.

**Comment by artemgur** [comment]
I found a simple workaround.
You can import the same objects from `keras.api` and its submodules. Pylance correctly works with such imports, and intellisense fully works.
For example, replace `from keras.optimizers import SGD` with `from keras.api.optimizers import SGD`

**Comment by AlanBogarin** [comment]
I'm sorry I wasn't there when this was open, as I had the same problem during the development of my packages.

In short, the reason pylance doesn't recognize exported symbols is because it doesn't follow the Python convention for public symbol interfaces.

you can read it [here](https://typing.python.org/en/latest/spec/distributing.html#library-interface-public-and-private-symbols)

To solve this you have two ways to do it

1. create the `__all__` symbol that will contain all the public symbols of the module or package
```python
# keras/__init__.py
from ... import activations, applications, ...

__all__ = (
    "activations",
    "applications",
    "backend",
    ...  # all the symbols that will be public
)
```
2. Imports should use a redundant alias (this tells type checkers to treat them as public, since they are private by default)
```python
from ... import activations as activations
from ... import applications as applications
...
```

**Comment by rivershah** [comment]
This api/src division is causing problems and breaks conventions. It breaks IDE autocompletion and import resolution and forces developers to implement hacks or use non-standard imports and creates unnecessary confusion

`pytorch`, `numpy` and other major libraries separate implementation from API without breaking standard import patterns.
Can we please consider reverting to a conventional package structure so all linting and analysis tools just work out of the box. 

These kinds of imports should work seamless in vscode and other ides but don't:

`from keras.layers import Dense`

**Comment by AlanBogarin** [comment]
This problem can be solved not by changing the structure, but by creating a script that fixes the formatting of modules generated by `namex`.

But I also think it's a good opportunity to significantly reduce the package size, considering @rivershah 's suggestion.

I opened an issue in another repository that appears to be a version compatible with type checking, so I became quite interested and reported this same problem with the `src/api` structure.

https://github.com/keras-team/keras-rs/issues/38

**Comment by rivershah** [comment]
Maybe I am asking for too big of refactor after all and I don't understand the context in which the decision for the api / src split was made.

I am totally ok with @AlanBogarin suggestion as long as downstream tools work nicely. As an end user, I just want to be able to:

```
from keras.models import Sequential
from keras.layers import Dense
```

And not get hit with a bunch of warnings / lack of important intellisense support.

**Comment by rivershah** [comment]
Can a maintainer with sufficient privileges please reopen issue

**Comment by mattdangerw** [comment]
Yeah let's fix this, this is pretty broken. `from keras import layers` needs to not appear broken with typecheckers.

@AlanBogarin thanks for the note, I think the redundant alias is probably a good fit for us.

I also think we need should ditch the `__path__` shenanigans at the top level init. We *do* want a separate exported api structure (this is important). Ideally, we don't want the toplevel browsing experience of our github repo to be a bunch of empty directories and init files (nice to have, not essentially). But I think we can get everything we want with standard packaging tools (setuptools `package_dirs`) so the packaged version of Keras does not the `api/` redirection and the public modules are where you'd expect.

I'll work on proposing a change with the redundant alias suggestion. Definitely we can find something that keeps our explicit api surface, and satisfies the typecheckers.

**Comment by mattdangerw** [comment]
Ok, here's a possible fix. Added the redudant export names, though I think the big issue might be the `__path__` hacking in the package `__init__` that would trip up IDEs and the like -- that's gone too.

The `api/` structure is now purely on the github site, package side we have.

```python
keras
├── activations
│   └── __init__.py
....  # Public init files with redundant imports.
├── src
....  # Private stuff subject to API changes.
```

If folks are able to test this out much appreciated! Particular with typechecking and development setups you are using. I pushed a package to a `keras-test` name on pypi.

Any of these (probably all good to check):
`pip install keras-test`
`pip install git+https://github.com/mattdangerw/keras@fix-package`
`!git clone https://github.com/mattdangerw/keras -b fix-package && pip install -e keras`

https://github.com/keras-team/keras/compare/master...mattdangerw:fix-package

**Comment by AlanBogarin** [comment]
Following the approach of maintaining the src/api structure, consider that extending the `namex` functionality so that when generating modules it creates them correctly is a better option, since it adds little code.

**Comment by mattdangerw** [comment]
> Following the approach of maintaining the src/api structure, consider that extending the namex functionality so that when generating modules it creates them correctly is a better option, since it adds little code.

Oh yes, sorry did that too. It won't save us the massive diff here, since we check in the generated api structure. But it's here, if we go with this approach.

https://github.com/fchollet/namex/compare/main...mattdangerw:redudant-imports?expand=1

**Comment by mattdangerw** [comment]
@rivershah @AlanBogarin and anyone else opened PRs for this. Can you please try these out and let us know if there is any lingering issues still? Much appreciated!

https://github.com/keras-team/keras/pull/21187
https://github.com/fchollet/namex/pull/6

**Comment by rivershah** [comment]
@mattdangerw Thanks, works very well and tested with `pylance`, `jedi`  and `ruff`.

**Comment by mattdangerw** [comment]
@rivershah Functional is not actual an exported symbol in Keras. The general approach is use `keras.Model(inputs, outputs)`. We have talked about opening it up, but only if we have a clear use for it. Why do you need it?

**Comment by rivershah** [comment]
A `keras2` hangup. Refactored in downstream; all fixed and  we don't need to export. No further clear reason to considering `keras.Model` does the trick.

**Comment by rivershah** [comment]
@mattdangerw I suspect the same issue is also present in `keras_hub` based models / layers. I can open a seperate ticket if you'd like for that repo.

For example type checkers complain about:
`from keras_hub.models.gemma3 import Gemma3Backbone`

**Comment by mattdangerw** [comment]
@rivershah indeed. Just made a similar PR in KerasHub. https://github.com/keras-team/keras-hub/pull/2222

These should both be available via nightly packages. `pip install keras-nightly` and `pip install keras-hub-nightly`. I think it probably makes sense to give these a few days and see if we get any bugs, but we can release these soon.
